### PR TITLE
options/ansi: Implement several of the fenv family functions

### PR DESCRIPTION
**Pull request by Dennisbonke**
Chromium made extensive use of some of these functions, so I decided to implement them like musl does. Fixes some ensures in Chromium.

Part of the Chromium on Managarm project.

### CODE REVIEW

File: options/ansi/generic/fenv-stubs.cpp [hunk 1]
```diff
@@ -2,54 +2,45 @@
...
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
-
-int fegetenv(fenv_t *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+// The functions that are not in this file but are defined in the header
+// are implemented like musl does in assembly.
+#ifdef __cplusplus
```
Comment (Geertiebear) [inline options/ansi/generic/fenv-stubs.cpp:7]: Why this ifdef?

**Review by Geertiebear — CHANGES_REQUESTED** [review]
These functions should be tested.

**Comment by no92** [comment]
I'm unclear on how these should be tested, TBH, and given that these are taken from musl (and have already been tested with bionic's tests), I don't see _that_ great of a need to do so.

**Comment by Dennisbonke** [comment]
I agree with Leo here. What do you guys think?

**Comment by avdgrinten** [comment]
Why is it hard to test them? Isn't it enough to cause an FPU exception and see if it behaves as expected?

**Comment by ArsenArsen** [comment]
in process floating point tests seem like the easiest thing a libc could be testing for

File: options/ansi/generic/fenv-stubs.cpp [hunk 2]
```diff
@@ -2,54 +2,39 @@
...
-	__builtin_unreachable();
-}
-
-int fegetenv(fenv_t *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// The functions that are not in this file but are defined in the header
+// are implemented like musl does in assembly.
+extern "C" __attribute__((__visibility__("hidden"))) int __fesetround(int);
```
Comment (64) [inline options/ansi/generic/fenv-stubs.cpp:7]: Isn't making the assembly hidden sufficient? Also, you could use C++-style attributes

File: options/ansi/include/bits/ansi/fenv.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#ifndef MLIBC_FENV
```
Comment (64) [inline options/ansi/include/bits/ansi/fenv.h:1]: _H

File: options/ansi/include/bits/ansi/fenv.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#define FE_TONEAREST 0
+#define FE_TOWARDZERO 1
+#define FE_DOWNWARD 2
+#define FE_UPWARD 3
+
+#else
+#error Unknown architecture
+#endif
+
+#endif // MLIBC_FENV
```
Comment (64) [inline options/ansi/include/bits/ansi/fenv.h:53]: _H

File: tests/ansi/fenv.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <fenv.h>
+#include <math.h>
+
+#define NO_OPTIMIZE(x) asm volatile("" :: "r,m" (x) : "memory")
```
Comment (64) [inline tests/ansi/fenv.c:6]: Dodgy, but I guess as long as the tests keep passing then who cares.

**Review by 64 — APPROVED** [review]

File: options/ansi/generic/fenv-stubs.cpp (hunk 2 above)
Comment (64) [reply options/ansi/generic/fenv-stubs.cpp:7]: Nvm, disregard. But C++ attribute would be nice

Comment (Dennisbonke) [reply options/ansi/generic/fenv-stubs.cpp:7]: Couldn't get it to work, so we can fix that later if need be.

**Pull request merged**